### PR TITLE
[Offload] Refactor offload test requirements

### PR DESCRIPTION
**Pull request by EthanLuisMcDonough**
Many tests in the `offload` project have requirements defined by which targets are not supported rather than which platforms are supported. This patch aims to streamline the requirement definitions by adding four new feature tags: `host`, `gpu`, `amdgpu`, and `nvidia-gpu`.

### CODE REVIEW

**Review by agozillon — APPROVED** [review]
LGTM I think the overall simplification is nice (at least from the perspective of someone writing these on a fairly regular basis at least), but please do wait on one of the more experienced reviewers to approve before landing! They'll likely know the broader impact of simplifying the architecture listing, if there is any :-)

File: offload/test/mapping/prelock.cpp
Comment (shiltian) [inline offload/test/mapping/prelock.cpp:5]: Can we also use `nvidia-gpu` here?

Comment (EthanLuisMcDonough) [reply offload/test/mapping/prelock.cpp:5]: Yes, I that should work. I kept those long names because I wasn't sure which targets this test was meant to run on (seeing that host, `nvptx64-nvidia-cuda`, `nvptx64-nvidia-cuda-LTO`, and `amdgcn-amd-amdhsa`). I figured it was best to err on the side of caution.

**Review by jhuber6 — APPROVED** [review]

File: offload/test/mapping/prelock.cpp
Comment (jhuber6) [inline offload/test/mapping/prelock.cpp:6]: Seems weird that we're doing `nvidia-gpu` instead of `nvptx64-nvidia-cuda` or `just `nvptx` when there's `amdgcn` right below it. I don't think there's a particular need to distinguish between the NVPTX triple and the NVPTX triple on a GPU.

Comment (EthanLuisMcDonough) [reply offload/test/mapping/prelock.cpp:6]: I'm just not sure where this test is supposed to be run. The test is unsupported on host targets, `amdgcn-amd-amdhsa`, and NVIDIA targets. Ideally we would only include the platform its meant to run on, but I'm not sure what platform that is.

Comment (EthanLuisMcDonough) [reply offload/test/mapping/prelock.cpp:6]: I'm just not sure where this test is supposed to be run. The test is unsupported on host targets, `amdgcn-amd-amdhsa`, and NVIDIA targets. Ideally we would only include the platform its meant to run on, but I'm not sure what platform that is.

Comment (EthanLuisMcDonough) [reply offload/test/mapping/prelock.cpp:6]: bug49334.cpp is another test case where I did this. I wasn't sure which targets the test was meant to support, so I just kept in a lot of the unsupported directives.

Upstream main:
https://github.com/llvm/llvm-project/blob/aa9bf71ca4eac50f4dfad9bb7b302491265cb889/offload/test/offloading/bug49334.cpp#L8-L16

My PR:
https://github.com/llvm/llvm-project/blob/88b8437ccfe02e9c4911610d3bcc31985f345918/offload/test/offloading/bug49334.cpp#L8-L11

Are these tests meant to be disabled?

Comment (jhuber6) [reply offload/test/mapping/prelock.cpp:6]: Lol, that list pretty much means "no one". I would just keep `nvptx64-nvidia-cuda` but drop the `-LTO` bit since it's irrelevant AFAIK. (I think there's a single test case where it matters but that's just because I don't have compiler-rt yet)

Comment (EthanLuisMcDonough) [reply offload/test/mapping/prelock.cpp:6]: I think the test where that matters is assert.c:

https://github.com/llvm/llvm-project/blob/88b8437ccfe02e9c4911610d3bcc31985f345918/offload/test/libc/assert.c#L6-L8

If the test isn't meant to run anywhere yet, I think `// UNSUPPORTED: host, gpu` might be more explicit.

Comment (jhuber6) [reply offload/test/mapping/prelock.cpp:6]: Oh yeah I wanted to get rid of that as well and just make it use the `libc` one, but for now it's probably fin. Can probably just have `REQUIRES: lto` or something.

**Pull request merged**
